src/shared_utilities/utils.jl

```diff
@@ -265,6 +267,75 @@ function DriverUpdateCallback(updateat::Vector{FT}, updatefunc) where {FT}
     )
 end
 
+"""
+    CheckpointCallback(checkpoint_frequency::Union{AbstractFloat, Dates.Period},
+                        output_dir, start_date, t_start; model, dt)
+
+Constructs a DiscreteCallback which saves the state to disk with the
+`save_checkpoint` function.
+
+# Arguments
+- `checkpoint_frequency`: The frequency at which checkpoints are saved. Can be
+  specified as a float (in seconds) or a `Dates.Period`.
+- `output_dir`: The directory where the checkpoint files will be saved.
+- `start_date`: The start date of the simulation.
+- `t_start`: The starting time of the simulation (in seconds).
+- `model`: The ClimaLand model object.
+- `dt`: The timestep of the model (optional), used to check for consistency.
+
+The callback uses `ClimaDiagnostics.EveryCalendarDtSchedule` to determine when
+to save checkpoints based on the `checkpoint_frequency`. The schedule is
+initialized with the `start_date` and `t_start` to ensure that the first
+checkpoint is saved at the correct time.
+
+The `save_checkpoint` function is called with the current state vector `u`, the
+current time `t`, and the `output_dir` to save the checkpoint to disk.
+"""
+function CheckpointCallback(
+    checkpoint_frequency::Union{AbstractFloat, Dates.Period},
+    output_dir,
+    start_date,
+    t_start;
+    model,
+    dt = nothing,
+)
+    # TODO: Move to a more general callback system. For the time being, we use
+    # the ClimaDiagnostics one because it is flexible and it supports calendar
+    # dates.
+
+    if checkpoint_frequency isa AbstractFloat
+        # Assume it is in seconds, but go through Millisecond to support
+        # fractional seconds
+        checkpoint_frequency_period =
+            Dates.Millisecond(1000checkpoint_frequency)
+    else
+        checkpoint_frequency_period = checkpoint_frequency
+    end
+
+    schedule = EveryCalendarDtSchedule(
+        checkpoint_frequency_period;
+        reference_date = start_date,
+        date_last = start_date + Dates.Millisecond(1000t_start),
+    )
+
+    if !isnothing(dt)
+        dt_period = Dates.Millisecond(1000dt)
+        if !isivisible(checkpoint_frequency_period / dt_period)
+            @warn "Checkpoint frequency ($(checkpoint_frequency_period)) is not an integer multiple of dt $(dt_period)"
+        end
+    end
+
+    cond = let schedule = schedule
```

Comment:
Where is the schedule function defined? Im confused because on line 315, it is a type. I guess all types are functions, but what is the default behavior?